src/games/seabattle/ai.ts

```diff
@@ -0,0 +1,191 @@
+import { IAIConfig } from '../index';
+import { ICell, IShip, ISalvo, generateRandomShips } from './game';
+
+interface IAISalvo extends ISalvo {
+  up: boolean;
+  down: boolean;
+  left: boolean;
+  right: boolean;
+}
+
+interface IPlayState {
+  G: any;
+  ctx: any;
+}
+
+class SeabattleBot {
+  async play(state: IPlayState, playerID: string) {
+    if (state.ctx.phase === 'setup') {
+      const shipPositions = generateRandomShips(1);
+      return this.makeSetShipsMove(shipPositions);
+    } else {
+      const cell = this.generateMove(state);
+      return this.makeSalvoMove(cell);
+    }
+  }
+
+  makeSetShipsMove(ships: IShip[]) {
+    return { action: { type: 'MAKE_MOVE', payload: { type: 'setShips', args: [ships], playerID: '1' } } };
+  }
+
+  makeSalvoMove(cell: ICell) {
+    return { action: { type: 'MAKE_MOVE', payload: { type: 'salvo', args: [cell.x, cell.y], playerID: '1' } } };
+  }
+
+  generateMove(state: IPlayState) {
+    const salvos = state.G.salvos.filter((salvo: ISalvo) => (
+      salvo.player === 1 &&
+      salvo.hit === true &&
+      typeof salvo.hitShip !== 'undefined'));
+    for (const salvo of salvos) {
+      // if we have already fired salvos, see if we've hit any ships that remain unsunk
+      if (!this.isShipSunk(state, salvo.hitShip)) {
+        const otherHitSalvos = this.getOtherSalvosHitShip(state, salvo.hitShip);
+        if (otherHitSalvos.length >= 2) {
+          // we have hit this ship at least twice, so we can make an intelligent move
+          return this.getMoveFromOtherSalvos(state, otherHitSalvos);
+        }
+        // no other salvos for the same ship were hit, so hit a random neighbor
+        return this.getRandomNeighbor(state, salvo);
+      }
+    }
+    // generate a random move
+    return this.generateRandomMove(state);
+  }
+
+  getMoveFromOtherSalvos(state: IPlayState, hitSalvos: ISalvo[]): ICell {
+    const len = hitSalvos.length;
+    const lastHitSalvo = hitSalvos[len - 1];
+    const lastHitCell = lastHitSalvo.cell;
+    const secondLastHitCell = hitSalvos[len - 2].cell;  // second to last hit
```

Comment:
I dont like that you are relying on the order of the salvos, doesnt make sense to me.
Why would you instead do this:
```
minSalvoPos = { x: (min of all x in hitSalvos), y: (min of all y in hitSalvos) }
maxSalvoPos = { x: (max of all x in hitSalvos), y: (max of all y in hitSalvos) }
diffMinMaxSalvoPos = { x: minSalvoPos.x - maxSalvoPos.x, y: minSalvoPos.y - maxSalvoPos.y }  // either x or y must be 0, given that the ship must be horizontal or vertical
direction = diffMinMaxSalvoPos.x === 0 ? { x: 0, y: 1 } :  { x: 1, y: 0 };
diffLength = diffMinMaxSalvoPos.x + diffMinMaxSalvoPos.y
if (diffLength == salvos.length) { // This means that there is no "hole" in the salvos, therefore we must try the edges
  return anyValidMove([{ x: minSalvoPos.x - direction.x, y: minSalvoPos.y - direction.y }, { x: maxSalvoPos.x + direction.x, y: maxSalvoPos.y + direction.y }]); 
} else {
  return anyValidMove([ all cells between minSalvoPos and maxSalvoPos ]);
}
```